File: mirror/tier2-php/showcase/type_basics.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: Type System
// Module: type_basics.php
// Purpose: Demonstrate PHP type system basics

function check_type($value): string {
    if ($value === null) return 'NULL';
    if (is_bool($value)) return 'boolean';
    if (is_int($value)) return 'integer';
    if (is_float($value)) return 'double';
    if (is_string($value)) return 'string';
    if (is_array($value)) return 'array';
    return 'object';
}

function is_numeric_type($value): bool {
    return is_int($value) || is_float($value);
}

function is_sequence($value): bool {
    return is_array($value) || is_string($value);
}

function combine_types($a, $b): string {
    if (is_string($a) && is_string($b)) {
        return $a . $b;
    } elseif (is_numeric_type($a) && is_numeric_type($b)) {
        return (string)($a + $b);
    }
    return 'mixed';
}

$result = [
    'type_check_int' => check_type(42),
    'type_check_str' => check_type('hello'),
    'is_numeric_int' => is_numeric_type(100),
    'is_numeric_str' => is_numeric_type('text'),
    'is_sequence_list' => is_sequence([1, 2, 3]),
    'combined_strings' => combine_types('hello', 'world'),
    'combined_numbers' => combine_types(10, 20),
];

echo json_encode($result);
?>

File: mirror/tier2-dart/showcase/type_basics.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Type System
// Module: type_basics.dart

String checkType(dynamic value) {
  if (value == null) return 'null';
  if (value is bool) return 'bool';
  if (value is int) return 'int';
  if (value is double) return 'double';
  if (value is String) return 'String';
  if (value is List) return 'List';
  if (value is Map) return 'Map';
  return 'dynamic';
}

bool isNumericType(dynamic value) {
  return value is int || value is double;
}

bool isSequenceType(dynamic value) {
  return value is List || value is String;
}

dynamic combineTypes(dynamic a, dynamic b) {
  if (a is String && b is String) return a + b;
  if (isNumericType(a) && isNumericType(b)) return a + b;
  return 'mixed';
}

void main() {
  final result = <String, dynamic>{
    'typeCheckInt': checkType(42),
    'typeCheckStr': checkType('hello'),
    'isNumericInt': isNumericType(100),
    'isNumericStr': isNumericType('text'),
    'isSequenceList': isSequenceType([1, 2, 3]),
    'combinedStrings': combineTypes('hello', 'world'),
    'combinedNumbers': combineTypes(10, 20),
  };
  print(result);
}
